Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,209 @@
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `host_context_handle` - output parameter. On success receives an opaque value which identifies the initialized host context. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can be called multiple times in a process.
+* If it's called when no runtime is present, it will run through the steps to "initialize" the runtime (resolving frameworks and so on).
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
+
+It needs to be possible to call this function simultaneously from multiple threads at the same time as well.
+It also needs to be possible to call this function while there is an active host context created by `hostfxr_initialize_for_app` and running inside the `hostfxr_run_app`.
+
+*TODO: What is the indication that there is no runtime in the process yet?*
+
+
+### Inspect and modify host context
+
+#### Runtime properties
+``` C++
+int hostfxr_get_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    char_t * value_buffer,
+    size_t value_buffer_size,
+    size_t * value_buffer_used);
+
+int hostfxr_set_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    const char_t * value);
+```
+
+These functions allow the native host to inspect and modify runtime properties. After initialization the properties should contain everything which the hosting components calculates. This gives the native host the opportunity to inspect, modify and resolve any potential conflicts.
+* `host_context_handle` - the initialized host context.
+* `name` - the name of the runtime property to get/set. Must not be `nullptr`.
+* `value` - the value of the property to set. When set to `nullptr` the property is "unset" - removed from the runtime property collection.
+* `value_buffer` - buffer to receive the value of the property
+* `value_buffer_size` - the size of the `value_buffer` is `char_t` units.
+* `value_buffer_used` - when successful contains number of `char_t` units used in the buffer (including the null terminator). If the specified `value_buffer_size` was too small, the function returns `HostApiBufferTooSmall` and sets the `value_buffer_used` to the minimum required size.
+
+Trying to get a property which doesn't exist is an error and `hostfxr_get_runtime_property` will return appropriate error code.
+
+Setting properties is only supported when the host context was initialized without any runtime in the process. This is really a limitation of the runtime for which the runtime properties are immutable.
```

Comment:
Yes, only the first initialized host context will allow to set property values. I changed the wording and added some explanation. In theory there's a small window where the first host context is initialized but runtime has not started yet, where it could be possible to change properties even from other host contexts, but it's not worth the complexity. Whatever mechanism we choose to communicate which context was the first will have to be thread safe, so the native host should base its behavior on that flag.

Additional runtime configs specifying properties after the first context has been initialize - that is a problem. I'm not sure what to do here.
* On one hand it would be safer to simply fail - either on any property, or if the property has different value (comparing the values is a bit tricky, we would have to do case sensitive string match, but for example file paths still might be equal even if we don't recognize them as such...). There's also the added complexity of defaults, where if the property is not set, it typically has some kind of default value, but it can be argued that in that case the runtime config "doesn't care".
* On the other hand I can easily imagine cases where people using 3rd party components will want to load them regardless of runtime config properties. For example if the component for some super weird reason says that it wants server GC, it will very likely work just fine on workstation GC, but strict property comparison would fail.

Just thinking out loud:
* Maybe we could "ignore but notify"  - somehow keep the properties on the second context and return a special error code. It would be still possible to use the context to "run", with the expectation that the property values are ignored. 
* At the same time though we should somehow "merge" the properties from the first context.
* Maybe if the properties match - merge them so that in the "success" case the native host can inspect properties even on the second host context and get a correct and full picture of what the runtime sees.
* And if they don't match - return the special error code and don't merge, in which case the host would only see the properties coming from the second context - and thus be able to decide if it wants to "ignore" them and run anyway or not.
* Would be relatively complex behavior, but it might be the best of both worlds...

Reply:
The ignore/notify and give the host a chance to decide behaviour sounds reasonable to me. I'd agree that it would be the best of both worlds.